File: 00_SHARE/01_past_questions/01_QE/22H2/22H2_Q2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>


using namespace std;

class GNode {
public:
    string id;

    GNode(string id) : id(id) {}

    string getId() const {
        return id;
    }

    // 비교 연산자 정의
    bool operator==(const GNode& other) const {
        return id == other.id;
    }
};

// GNode의 해시 함수 정의
namespace std {
    template <>
    struct hash<GNode> {
        size_t operator()(const GNode& node) const {
            return hash<string>()(node.id);
        }
    };
}

vector<vector<string>> paths(unordered_map<GNode, vector<GNode>>& G, GNode s, GNode t) {
    vector<vector<string>> result;

    function<void(GNode, vector<string>)> dfs = [&](GNode curr, vector<string> path) {
        if (curr.getId() == t.getId()) {
            path.push_back(curr.getId());
            result.push_back(path);
        } else {
            path.push_back(curr.getId());
            for (GNode neighbor : G[curr]) {
                dfs(neighbor, path);
            }
        }
    };

    dfs(s, {});

    return result;
}

int main() {
    GNode a("a");
    GNode b("b");
    GNode c("c");
    GNode d("d");

    unordered_map<GNode, vector<GNode>> G;
    G[a] = {b, c};
    G[b] = {d};
    G[c] = {};
    G[d] = {c};

    vector<vector<string>> result = paths(G, a, c);

    cout << "Paths from 'a' to 'c':" << endl;
    for (vector<string> path : result) {
        for (string node : path) {
            cout << node << " ";
        }
        cout << endl;
    }

    return 0;
}

File: 00_SHARE/01_past_questions/01_QE/22H2/22H2_Q2.py
## Q2 ##
class GNode:
  def __init__(self, id):
    self.id = id
  
  def __str__(self):
    return self.id
  
def paths(G, s, t):
    paths = []

    def dfs(curr, path):
        #현재 노드가 목표 노드인 경우
        if curr.id == t.id:
            #경로를 paths 리스트에 추가한다
            paths.append(path + [curr.id])
        else:
            #현재 노드의 이웃들에 대해 이를 반복
            for neighbor in G[curr]:
                dfs(neighbor, path + [curr.id])
    
    dfs(s, [])
    return paths

# Driver Code
if __name__ == '__main__':

    a, b, c, d = GNode('a'), GNode('b'), GNode('c'), GNode('d')
    G = dict()
    G[a], G[b], G[c], G[d] = [b, c], [d], [], [c]

    print(paths(G, a, c)) # [['a', 'b', 'd', 'c'], ['a', 'c']]
